sharedingress-config-generator/controller.go

```diff
@@ -0,0 +1,173 @@
+package sharedingressconfiggenerator
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"path/filepath"
+	"time"
+
+	hyperv1 "github.com/openshift/hypershift/api/hypershift/v1beta1"
+	"github.com/openshift/hypershift/support/util"
+
+	routev1 "github.com/openshift/api/route/v1"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type SharedIngressConfigReconciler struct {
+	Client client.Client
+
+	ConfigPath               string
+	HAProxyRuntimeSocketPath string
+
+	// lastReloadedHash stores the hash of the last successfully reloaded configuration
+	lastReloadedHash []byte
+}
+
+func (r *SharedIngressConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.Client = mgr.GetClient()
+
+	// A channel is used to generate an initial sync event.
+	// Afterwards, the controller syncs on HostedClusters.
+	initialSync := make(chan event.GenericEvent)
+
+	go func() {
```

Comment:
why did we need the initial sync?

Reply:
to generate the initial config file at start. Otherwise we won't generate anything until a HostedCluster is created and the HAProxy will crash loop because of the missing file.